core/src/cpu/event.rs

```diff
@@ -1,7 +1,64 @@
+use std::array;
+
 use serde::{Deserialize, Serialize};
 
 use crate::runtime::Instruction;
+use crate::runtime::LookupIdSampler;
 use crate::runtime::MemoryRecordEnum;
+use crate::runtime::Opcode;
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub enum LookupIds {
+    AluLookupId(u128),
+    SyscallLookupId(u128),
+    MemoryLookupIds([u128; 2]),
+    BranchLookupIds([u128; 3]),
+    JumpLookupIds([u128; 2]),
+    AuipcLookupId(u128),
+
+    /// Used for the case of value of no lookupids.
+    DefaultLookupIds,
+}
+
+impl LookupIds {
+    pub fn new(instr: Instruction, rng_sampler: &mut impl LookupIdSampler) -> Self {
+        let num_lookup_ids = if instr.is_alu_instruction()
+            || instr.is_ecall_instruction()
+            || instr.opcode == Opcode::AUIPC
+        {
+            1
+        } else if instr.is_branch_instruction() {
+            3
+        } else if instr.is_jump_instruction() || instr.is_memory_instruction() {
+            2
+        } else {
+            0
+        };
+
+        let lookup_ids = rng_sampler.sample(num_lookup_ids);
+
+        if instr.is_alu_instruction() {
+            LookupIds::AluLookupId(lookup_ids[0])
+        } else if instr.is_ecall_instruction() {
+            LookupIds::SyscallLookupId(lookup_ids[0])
+        } else if instr.is_memory_instruction() {
+            LookupIds::MemoryLookupIds([lookup_ids[0], lookup_ids[1]])
+        } else if instr.is_branch_instruction() {
+            LookupIds::BranchLookupIds([lookup_ids[0], lookup_ids[1], lookup_ids[2]])
+        } else if instr.is_jump_instruction() {
+            LookupIds::JumpLookupIds([lookup_ids[0], lookup_ids[1]])
+        } else if instr.opcode == Opcode::AUIPC {
+            LookupIds::AuipcLookupId(lookup_ids[0])
+        } else {
+            LookupIds::DefaultLookupIds
+        }
+    }
+
+    pub fn new_sublookups(rng_sampler: &mut impl LookupIdSampler) -> [u128; 6] {
```

Comment:
what's the point of having this enum if we're just going to use [u128; 6]?

shouldn't we do the same for Alu lookups?

Reply:
Ya, this shouldn't be part of the enum function.  I didn't want to add a variant with [u128; 6] internal type, since that would made the size of CpuLookupIds much bigger.

Reply:
The reason I didn't do it for ALU lookups is that it seems like all 6 of the sub_lookup ids are needed for the divrem chip.

I was considering making the sublookups optional, but there will be no space savings (an Option type is as big as the internal type).

However, it would save on the time to generate random ids, so its probably worth making it optional.